File: sgsort.py
# sgsort.py
# Author: George Lesica <glesica@gmail.com>
# Description: A Python implementation of Shotgun Sort.

import random


def sgsorted(iterable, reverse=False):
    """
    Accepts an iterable `iterable` and returns a copy of 
    it that has been sorted in ascending order.

    If `reverse` is True, descending order will be used.
    """
    def is_sorted(iterable):
        a = iterable[0]
        for i in iterable:
            if i < a:
                return False
            a = i
        return True

    l = iterable[:] # Makes a copy of iterable

    while (not is_sorted(l)):
        random.shuffle(l)

    if reverse:
        l.reverse()

    return l


if __name__ == '__main__':
    l = [3,2,4,1,5]
    print l
    print sgsorted(l)

    l = [5,4,6,3,7,2,9,8,1]
    print l
    print sgsorted(l)

File: sgsort.c
/*
 * shotgun.c
 * Author: George Lesica <george@lesica.com>
 * Description: An ANSI C implementation of the "shotgun sort" algorithm.
 *
 */

#include <stdlib.h>
#include <string.h>

/* Private utility functions. */

int sorted(
    const void *base,
    size_t num,
    size_t size,
    int (*comparator) (const void *, const void *)
) {
    int i;

    char *b = (char *) base;

    for (i=1; i<num; i++) {
        if (comparator(&b[(i-1)*size], &b[i*size]) > 0) {
            return 0;
        }
    }
    return 1;
}

void randomize(
    void *base,
    size_t num,
    size_t size
) {
    size_t i, j;
    char *b = (char *) base;
    void *temp = malloc(size);

    for (i=0; i<num; i++) {
        j = rand() % num;
        memcpy(temp, &b[i*size], size);
        memcpy(&b[i*size], &b[j*size], size);
        memcpy(&b[j*size], temp, size);
    }

    free(temp);
}

/* Public Functions */

int int_compare(const void *a, const void *b) {
    return (*(int *) a - *(int *) b);
}

void sgsort(
    void *base, 
    size_t num, 
    size_t size, 
    int (*comparator) (const void *, const void *)
) {
    /* Loop until the array is sorted */
    while (! sorted(base, num, size, comparator)) {
        randomize(base, num, size);
    }

}

